File: app/src/main/java/net/squanchy/schedule/view/OtherEventItemView.java
package net.squanchy.schedule.view;

import android.content.Context;
import android.support.annotation.DrawableRes;
import android.util.AttributeSet;
import android.widget.ImageView;
import android.widget.TextView;

import net.squanchy.R;
import net.squanchy.schedule.domain.view.Event;

import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;

public class OtherEventItemView extends EventItemView {

    private static final int NO_DRAWABLE = 0;
    
    private TextView titleView;
    private TextView timestampView;
    private ImageView illustrationView;

    public OtherEventItemView(Context context, AttributeSet attrs) {
        this(context, attrs, R.attr.cardViewDefaultStyle);
    }

    public OtherEventItemView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();

        timestampView = (TextView) findViewById(R.id.timestamp);
        titleView = (TextView) findViewById(R.id.title);
        illustrationView = (ImageView) findViewById(R.id.illustration);
    }

    @Override
    public void updateWith(Event event) {
        ensureSupportedType(event.type());

        timestampView.setText(startTimeAsFormattedString(event));
        titleView.setText(event.title());

        illustrationView.setImageResource(illustrationFor(event.type()));
    }

    private void ensureSupportedType(Event.Type type) {
        if (type == Event.Type.COFFEE_BREAK
                || type == Event.Type.LUNCH
                || type == Event.Type.OTHER
                || type == Event.Type.REGISTRATION
                || type == Event.Type.SOCIAL) {
            return;
        }
        throw new IllegalArgumentException("Event with type " + type.name() + " is not supported by this view");
    }

    private String startTimeAsFormattedString(Event event) {
        DateTimeFormatter formatter = DateTimeFormat.shortTime()
                .withZone(event.timeZone());

        return formatter.print(event.startTime().toDateTime());
    }

    @DrawableRes
    private int illustrationFor(Event.Type type) {
        switch (type) {
            case COFFEE_BREAK:
                return R.drawable.coffee_break;
            case LUNCH:
                return R.drawable.illustration_lunch;
            case REGISTRATION:
                return R.drawable.registration;
            case OTHER:
                return NO_DRAWABLE;
            case SOCIAL:
                return R.drawable.illustration_lunch;        // TODO replace these with the correct images once we have them
            default:
                throw new IllegalArgumentException("Type not supported: " + type.name());
        }
    }
}

File: app/src/main/java/net/squanchy/schedule/view/OtherEventItemView.kt
package net.squanchy.schedule.view

import android.content.Context
import android.util.AttributeSet
import androidx.annotation.DrawableRes
import androidx.core.view.isVisible
import kotlinx.android.synthetic.main.item_schedule_event_other.view.*
import net.squanchy.R
import net.squanchy.schedule.domain.view.Event
import net.squanchy.support.time.shortTimeFormatter

class OtherEventItemView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet,
    defStyleAttr: Int = R.attr.cardViewDefaultStyle
) : EventItemView(context, attrs, defStyleAttr) {

    override fun updateWith(event: Event, showRoom: Boolean, showDay: Boolean, showFavorite: Boolean) {
        event.type.ensureSupported()

        timestamp.text = startTimeAsFormattedString(event)
        title.text = event.title

        if (event.type != Event.Type.OTHER) {
            illustration.isVisible = true
            illustration.setImageResource(illustrationFor(event.type))
        } else {
            illustration.isVisible = false
            illustration.setImageDrawable(null)
        }
    }

    private fun Event.Type.ensureSupported() {
        when (this) {
            Event.Type.COFFEE_BREAK,
            Event.Type.LUNCH,
            Event.Type.OTHER,
            Event.Type.REGISTRATION,
            Event.Type.SOCIAL -> return

            else -> throw IllegalArgumentException("Event with type $name is not supported by this view")
        }
    }

    private fun startTimeAsFormattedString(event: Event): String {
        val timeZone = event.timeZone
        val formatter = shortTimeFormatter()
            .withZone(timeZone)

        return formatter.format(event.startTime.atZone(timeZone))
    }

    @DrawableRes
    private fun illustrationFor(type: Event.Type): Int = when (type) {
        Event.Type.COFFEE_BREAK -> R.drawable.coffee_break
        Event.Type.LUNCH -> R.drawable.lunch
        Event.Type.REGISTRATION -> R.drawable.registration
        Event.Type.SOCIAL -> R.drawable.social
        else -> throw IllegalArgumentException("Type not supported: ${type.name}")
    }
}
